tests/generators/cpp/src/test.cpp

```diff
@@ -0,0 +1,353 @@
+#include <type_traits>
+
+#include <catch2/catch.hpp>
+
+#if defined OPENASSETIO_TRAITGENTEST_INCLUDES_PACKAGE
+#include <openassetio_traitgen_test_all/openassetio_traitgen_test_all.hpp>
+#elif defined OPENASSETIO_TRAITGENTEST_INCLUDES_SUBPACKAGE
+#include <openassetio_traitgen_test_all/specifications/specifications.hpp>
+#include <openassetio_traitgen_test_all/traits/traits.hpp>
+#elif defined OPENASSETIO_TRAITGENTEST_INCLUDES_NAMESPACE
+#include <openassetio_traitgen_test_all/specifications/test.hpp>
+#include <openassetio_traitgen_test_all/traits/aNamespace.hpp>
+#include <openassetio_traitgen_test_all/traits/anotherNamespace.hpp>
+#else
+#error "An #include style must be chosen"
+#endif
+
+TEST_CASE("openassetio_traitgen_test_all - all expected traits are defined") {
+  using openassetio_traitgen_test_all::traits::aNamespace::NoPropertiesTrait;
+  STATIC_REQUIRE(std::is_class_v<NoPropertiesTrait>);
+  STATIC_REQUIRE(
+      std::is_base_of_v<openassetio::trait::TraitBase<NoPropertiesTrait>, NoPropertiesTrait>);
+  STATIC_REQUIRE(
+      std::is_class_v<
+          openassetio_traitgen_test_all::traits::aNamespace::NoPropertiesMultipleUsageTrait>);
+  STATIC_REQUIRE(
+      std::is_class_v<openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait>);
+  STATIC_REQUIRE(
+      std::is_class_v<openassetio_traitgen_test_all::traits::anotherNamespace::NoPropertiesTrait>);
+}
+
+TEST_CASE("openassetio_traitgen_test_all - traits have expected IDs") {
+  CHECK(openassetio_traitgen_test_all::traits::aNamespace::NoPropertiesTrait::kId ==
+        "openassetio-traitgen-test-all:aNamespace.NoProperties");
+  CHECK(openassetio_traitgen_test_all::traits::aNamespace::NoPropertiesMultipleUsageTrait::kId ==
+        "openassetio-traitgen-test-all:aNamespace.NoPropertiesMultipleUsage");
+  CHECK(openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait::kId ==
+        "openassetio-traitgen-test-all:aNamespace.AllProperties");
+  CHECK(openassetio_traitgen_test_all::traits::anotherNamespace::NoPropertiesTrait::kId ==
+        "openassetio-traitgen-test-all:anotherNamespace.NoProperties");
+}
+
+namespace {
+/**
+ * Fixture struct for parameterizing tests on trait property type.
+ *
+ * Uses knowledge of the expected structure of the generated
+ * AllPropertiesTrait.
+ *
+ * @tparam PropertyType Type of property to specialise for.
+ */
+template <typename PropertyType>
+struct PropertyFixture;
+
+template <>
+struct PropertyFixture<openassetio::Bool> {
+  using AllPropertiesTrait = openassetio_traitgen_test_all::traits::aNamespace::AllPropertiesTrait;
+
+  AllPropertiesTrait trait;
+
+  static inline const openassetio::trait::property::Key kPropertyKey = "boolProperty";
+  static constexpr std::string_view kPropertyType = "openassetio::Bool";
+  static constexpr openassetio::Bool kDefaultValue = false;
+  static constexpr openassetio::Bool kExpectedValue = true;
+  static constexpr openassetio::Int kUnexpectedValue = 123;
```

Comment:
[consider] rename to `kMismatchedTypedValue` or similar for clarity?

Reply:
Renamed in 13f0ac21